### ISSUE TITLE
Reconsider changes to pipe and compose

### ISSUE THREAD

**Issue by CrossEye** [ramda/ramda#1318]
In #1260, after a fairly tortuous path, we ended up reworking a number of the `composeX` / `pipeX` functions.  I'm afraid at this point that I believer we broke them, at lest the default `pipe`/`compose` versions.

Today, attempting to answer [this Stack Overflow question][so], I tried to use this code as a points-free alternative:

```js
// getLocalDeps :: [{ source :: { value :: String }}] -> RegExp -> [String]
const getLocalDeps =  R.useWith(
  R.flip(R.call),
  R.map(R.path(['source', 'value'])),
  R.pipe(R.test, R.filter)
);
```

I was stuck for quite a while trying to figure out [why it didn't work][v16].  Everything looks right.  Eventually I realized that `pipe` wasn't doing what it was supposed to.  It [works in the previous release][v15].  And of course I can [simply use a lambda instead][v17]:

```js
// getLocalDeps :: [{ source :: { value :: String }}] -> RegExp -> [String]
const getLocalDeps =  R.useWith(
  R.flip(R.call),
  R.map(R.path(['source', 'value'])),
  regex => R.filter(R.test(regex))
);
```
 or this:

```js
  // ...
  R.pipe(regex => item => R.test(regex, item), R.filter)
  //...
```

But those both miss some of the elegance of the original.  I really think that we want to be able to use `pipe`/`compose` in this manner.  This strikes me as a very common requirement.

  [so]: http://stackoverflow.com/questions/31701246/
  [v15]: http://bit.ly/1MtZhjB
  [v16]: http://bit.ly/1MtZfbs
  [v17]: http://bit.ly/1MtZhQz

**Comment by davidchambers** [comment]
I'd prefer to leave `R.compose` and `R.pipe` as they are now, and add `R.composeN` and `R.pipeN` to support functions with optional arguments and functions which misreport their arities. We do this for `R.curry` with `R.curryN`.

**Comment by asaf-romano** [comment]
Please also see #1314. 0.16+ is so far a no-go for me because many of my async pipelines happen to start with a sync function :-/

**Comment by CrossEye** [comment]
But note that this is neither of those.  I simply want to use `R.pipe(R.test, R.filter)`, pass it the regex and get back `[String] -> [String]`.  That worked before #1260.  It's not a matter of optional arguments or poorly reporting of arities.  I could also solve it for this problem with 

```js
  // ...
  R.pipe(R.unary(R.test), R.filter)
  // ...
```

But it feels really wrong to have to do that.  That's my problem here.  Maybe this will just end up being our go-to answer for these, but it really feels wrong.

**Comment by davidchambers** [comment]
> But note that this is neither of those.

It is and it isn't. `R.test`, like every non-unary function in Ramda, reports its arity correctly *and* incorrectly. The problem is that it's really two functions in one:

    R.test :: RegExp -> String -> Boolean
    R.test :: (RegExp,String) -> Boolean

There's no way for `R.pipe` to know which of these functions one wishes to use.

I'm beginning to wonder whether we should replace Ramda-flavoured currying with honest-to-goodness currying.

**Comment by CrossEye** [comment]
> I'm beginning to wonder whether we should replace Ramda-flavoured currying with honest-to-goodness currying.

Oh, damn, I've got Chambers thinking about further fundamental changes.  Uh oh!  :smile:

**Comment by davidchambers** [comment]
You have. At the moment we allow

```javascript
R.map(R.negate, [1, 2, 3])
```

as syntactic sugar for

```javascript
R.map(R.negate)([1, 2, 3])
```

This makes the library more familiar to JavaScript programmers, certainly. The question is what does this cost us? I need to give this some thought.

**Comment by paldepind** [comment]
> The question is what does this cost us? I need to give this some thought.

In many cases it makes point free definitions impossible. Take this function for instance, which I wrote earlier today (it's just a tiny wrapper around a recursive function that traverses a recursive tree data structure):

```javascript
const walkTreeByTasks = R.curry((tasks, nodes) => walkDown([], extractKeys(tasks), nodes))
```

Notice that `nodes` is the last argument and it also appears last in the function body. In Haskell this means we can make `nodes` point free. But in JavaScript where the function will be called like `walkTreeByKeys(keys, nodes)` we can't do that. In "honest-to-goodness curry"-land however it is not an issue (like in Haskell):

```javascript
const walkTreeByTasks = (tasks) => walkDown([])(extractKeys(tasks))
walkTreeByTasks(tasks)(nodes)
```

This has bothered me a bit here and there. But I really don't think going down that route is worth it compared to the obvious downsides.

I think the problem pointed out by @CrossEye here is quite severe. By the standard definition of compose this `R.compose(R.filter, R.test)(/^x/)` should be equivalent to `R.filter(R.test(/^x/))`. That is the _definition_ of what composing two functions means but we are breaking it :cry:

**Comment by davidchambers** [comment]
> By the standard definition of compose this `R.compose(R.filter, R.test)(/^x/)` should be equivalent to `R.filter(R.test(/^x/))`.

Good point, @paldepind! It also concerns me that `R.compose(f, R.identity)` is not equivalent to `R.compose(R.identity, f)` for non-unary values of `f`.

If we stick to the mathematic definition of composition then we can assume that all functions are unary and these problem go away.

**Comment by CrossEye** [comment]
@paldepind:

> In many cases it makes point free definitions impossible. Take this function for instance, which I wrote earlier today (it's just a tiny wrapper around a recursive function that traverses a recursive tree data structure):

> ```javascript
> const walkTreeByTasks = R.curry((tasks, nodes) => walkDown([], extractKeys(tasks), nodes))
> ```

Is `walkDown` curried?  And if so, would this work?:

```js
const walkTreeByTasks = R.useWith(walkDown([]), extractKeys, R.identity)
```

I'm not sure if it will.  It depends, I think, on how the first parameter to `walkDown` is used, but this is the sort of thing that `useWith` was written for.  It certainly is designed to overcome a limitation that I see in Ramda's `curry`/`pipe`.

But I'm definitely not ready to go the route David is suggesting.  I can't remember where the suggestion was, but I'm pretty sure it was David who listed a hierarchy of libraries / languages on a FP scale:  Underscore < Ramda < Sanctuary < Purescript.  I think a suggestion like this is trying to jump several levels on such a list, ignoring the purpose of Ramda.

**Comment by CrossEye** [comment]
@davidchambers:
>> By the standard definition of compose this `R.compose(R.filter, R.test)(/^x/)` should be equivalent to `R.filter(R.test(/^x/))`.

> Good point, @paldepind! 

Yes, a very good point.

> It also concerns me that `R.compose(f, R.identity)` is not equivalent to `R.compose(R.identity, f)` for non-unary values of `f`.

I think that's a lesser concern.  Certainly we don't expect `compose(f, g)` to be equivalent to `compose(g, f)` for a general case, so the real issue is more likely that `compose(f, identity)` is not equivalent to `f`.  That is true iff `f` is polyadic.  And it has nothing to do with our implementation.  `f(identity(x, ...args)); //=> f(x)`.  

But to say that we want the library to only deal with unary functions is a **huge** step away from familiarity for JS developers.  And I really don't think there is much of a gain.

**Comment by davidchambers** [comment]
> But to say that we want the library to only deal with unary functions is a **huge** step away from familiarity for JS developers.  And I really don't think there is much of a gain.

I'm not suggesting we assume all functions are unary throughout the library. I'm suggesting removing the special case for the first function in a composition. We already assume that functions beyond the first are unary; it doesn't strike me as a big step to require the first function to be unary as well.

**Comment by CrossEye** [comment]
> We already assume that functions beyond the first are unary; it doesn't strike me as a big step to require the first function to be unary as well.

But that means you can no longer use `compose` to build polyadic functions, which would be a big deal in my mind.  It reduces the power of `compose` and `pipe` significantly, and simply increases the need for functions like `useWith` and `converge`, which I've always wanted to find a way to gracefully remove.

To my mind `compose` is the very heart of Ramda, and saying that from now on it can only return unary functions would be to reduce Ramda's chief strength.

**Comment by paldepind** [comment]
@CrossEye 

Your `R.useWith` variant would work! It is a nice workaround way to achieve even more point-free-ness. But it is not as straightforward as doing it the other way.

> But I'm definitely not ready to go the route David is suggesting. I can't remember where the suggestion was, but I'm pretty sure it was David who listed a hierarchy of libraries / languages on a FP scale: Underscore < Ramda < Sanctuary < Purescript. I think a suggestion like this is trying to jump several levels on such a list, ignoring the purpose of Ramda.

Yes. I certainly agree on that which I also mentioned before. Demanding unary invocations would be very unnatural (given the syntax of JS) with only small benefits.

> But that means you can no longer use compose to build polyadic functions, which would be a big deal in my mind.

I agree. If such a restriction is put in place `compose` will always return a unary function. Then we are no better off (actually worse I think) than in the previous iteration where `compose` returned a non-curried function.

I'm currently in favor of removing the currying of the function created by `compose`. It's quite unfortunate but there really is some confusing ambiguity here.

Basically, in all cases where users want to have the first function partially applied and then passed to the next function in the chain we are making the wrong guess when we curry. It might be better to simply not guess at all and have users curry themselves or introduce a `composeN` for convenience.

Not currying functions created by `compose`/`pipe` will also be surprising to some users. But it is much less subtle. I think when even @CrossEye is making sentiments like: "I was stuck for quite a while trying to figure out why it didn't work. Everything looks right." Then the design is broken.

**Comment by CrossEye** [comment]
Ping.

Any further thoughts on this?  I will probably create a PR in a few days, but I know it's likely to be controversial.  I'd like to take as much into account as I can when I create it.

**Comment by davidchambers** [comment]
These are the functions I would like Ramda to provide:

    :: (b -> c) -> (a -> b) -> (a -> c)

    :: [(a -> b), (b -> c), …, (m -> n)] -> (a -> n)

    :: [((a,…) -> b), ((b,…) -> c), …, ((m,…) -> n)] -> (a -> … -> b -> … -> m -> … -> n)

The first is a curried binary function for right-to-left composition of two unary functions (any curried function can be considered unary). This could be named `compose`.

The second is a unary function for left-to-right composition of zero or more unary functions. This could be named `pipe`.

The third is my attempt at a type signature for the original behaviour of #1260, as depicted in https://github.com/ramda/ramda/pull/1260#issuecomment-120452772. It represents a unary function for left-to-right composition of zero or more functions of arity ≥ 1. The return value is a curried function of length `sum(map(length, fs)) - (fs.length - 1)`. I've no idea what we'd name this.

**Comment by CrossEye** [comment]
Thank you.  This is a very clean way to frame the issues.

The first signature sounds very much like what @buzzdecafe was suggesting at one point for a simple `compose`.  In his [gist][gi] he played around with an idea much like the third signature, and similar to what you initially attempted in #1260.  While those ideas interest me, I'm not particularly compelled by them.  If we could find a reasonable implementation, I'm sure we'd find a name that works.

I'm reluctant, but only slightly reluctant, to let go of our notion that `pipe` and `compose` are order-reversed twins.

But I'm much, much more reluctant to remove the ability of `pipe` and `compose` to return polyadic functions.

So these would feel much more comfortable to me:

```js
:: (b -> c) -> ((a,…) -> b) -> ((a,…) -> c)

:: [((a,…) -> b), (b -> c), …, (m -> n)] -> ((a,…) -> n)
```

I might phrase them differently, as the '`(a,…)`' syntax to me specifically implies a lack of currying:

```js
:: (m -> n) -> (a -> b -> … -> m) -> (a -> b -> …  -> n)

:: [((a -> b -> … -> m), (m -> n), …, (y -> z)] -> (a -> b -> …  -> z)
```
or some such.  But that's a detail right now.

Obviously the major difference is whether we want to support polyadic functions in the first-applied functions, and thus in the result of `compose` / `pipe`.  I think that's an important part of JS not to be lost.


  [gi]: https://gist.github.com/buzzdecafe/b00905f6487725173ff7

**Comment by davidchambers** [comment]
> Obviously the major difference is whether we want to support polyadic functions in the first-applied functions, and thus in the result of `compose` / `pipe`. I think that's an important part of JS not to be lost.

I agree that this is important. I was hoping the "third" function could serve this need, which would keep the other two functions simple.

**Comment by CrossEye** [comment]
The extra complexities of that third one scare me quite a bit.  I suppose if we overcome them, it could do this job (as well as quite a few others.)  But we'd have to overcome the problem you [boiled down][bd] to this simple example:

```js
> R.pipe(R.add, R.map)(10)([1, 2, 3])
[ 11, 12, 13 ]
```

I don't see a way to do that yet, but I still hold out hope.

  [bd: https://github.com/ramda/ramda/pull/1260#discussion_r34409228

**Comment by TheLudd** [comment]
@CrossEye 

> Any further thoughts on this?

Yes. As discussed in #732 (I am continuing the discussion here since this is the compose-thread)
With the aruments made by @aaditmshah in https://github.com/ramda/ramda/issues/732#issuecomment-133829133 I'd like to define compose as
```
var compose = curry(function(f, g, x) {
  return f(g(x))
})
```
We can also add `composeBinary` as proposed in that thread and other compose functions could be derived from them. For instance we could have `composeAll` which takes a list of functions and an argument and `pipe` could do the same. 

With the change to curry as proposed in #732 we could really get rid of `useWith` which we seem to agree on is currently a necessary evil or perhaps just evil.

**Comment by CrossEye** [comment]
https://gitter.im/ramda/ramda?at=55e22a1efbbdfcbd3b5ddfa2

**Comment by knubie** [comment]
I'm not sure if this is related but I ran into this situation today:
This code,
```javascript
compose(getAnyPieceAtPosition(board), getPosition(axis))(1);
```
returns a different result than this code.
```javascript
getAnyPieceAtPosition(board, getPosition(axis, 1))
```
I had assumed that compose would pass 1 to `getPosition` which would then pass its result to `getAnyPieceAtPosition`, but that doesn't seem to be the case.

This seems to be related to the issue in gitter.

**Comment by davidchambers** [comment]
The sequence of events is now clear to me.

1.  In the beginning, the children composed unary functions and all was well in the world.

2.  One day Alice wondered what would happen if she provided a binary function as the second argument to `R.compose`. Alice was pleased to discover that `R.compose(Math.sqrt, R.add)(9, 16)` gave `5` as she had hoped it would! All was well in the world.

3.  Months passed. While reworking a section of code Alice came across `R.compose(Math.sqrt, R.add)(9, 16)`. She realized it would be useful to partially apply the square-root-of-sum function. When she evaluated `R.compose(Math.sqrt, R.add)(9)` in the REPL, she was saddened to see `NaN` printed in her terminal.

4.  _Wait!_ Alice thought. _Many Ramda functions return curried functions. Surely `R.compose` could too._

5.  Weeks passed. Alice awoke early one morning to learn that a new version of Ramda had been published during the night. She scanned the upgrade guide and _Yes!_, `R.compose` had been updated to return a curried function. After breakfast Alice updated her __package.json__ and revisited the square-root-of-sum function which she was now able to partially apply. Once again, all was well in the world.

6.  That same day, Bob upgraded to the latest Ramda release. Bob was not happy. Lots of his tests were failing and he wasn't sure why. After nearly an hour of detective work, Bob discovered the problem. His code made heavy use of `R.pipe`, and in some cases the first function in the pipeline had optional arguments. _Who knew that `JSON.parse` is a **binary** function?_

7.  Bob spent much of the afternoon adding `R.identity` at the beginnings of dozens of pipelines, to foil those pesky functions with optional arguments.

This is where we currently stand: Alice is happy; Bob is not.

---

@CrossEye and I just had a fruitful discussion on Gitter. This matrix is acceptable to both of us:

| Name      | Arity | Arg arities   | Direction     | Returns curried |
| --------- |:-----:|:-------------:|:-------------:|:---------------:|
| `compose` |   2   |    (1,>=1)    | right-to-left |                 |
| `pipe`    |   1   | [>=1,1,1,...] | left-to-right |                 |
| `???`     |   1   | [>=1,>=1,...] | left-to-right |        ✔︎        |

Bob could then use `R.pipe` without fear: it would no longer return a curried function, so optional arguments would not cause problems.

Alice could opt in to currying by using the new function which composes any number of functions of arbitrary arity.

All could once again be well in the world.

**Comment by paldepind** [comment]
> That same day, Bob upgraded to the latest Ramda release. Bob was not happy. Lots of his tests were failing and he wasn't sure why. After nearly an hour of detective work, Bob discovered the problem. His code made heavy use of R.pipe, and in some cases the first function in the pipeline had optional arguments. Who knew that JSON.parse is a binary function?

That is definitely one problem. But optional arguments is not the biggest IMO. The biggest is that you can't do stuff like this:

```js
R.compose(R.flip(R.map)([1,2,3,4]), R.add)(3);
```

I.e. the current compose does not work whenever you want to partially apply the first function. That is also the issue @CrossEye had when he opened this issue.

Attempting to automatically curry the function returned by `compose` is too naive. If a function is only guessing the correct arity half the time then it's better to not guess at all.

To me the lack of a good name for the last proposed function is bad sign.

I prefer `pipe` and `compose` to be symmetric.

I think both `compose` and `pipe` should be variadic. Anything else seems to sacrifice API convenience for implementation convenience and for satisfying a theoretical aesthetic.

**Comment by davidchambers** [comment]
> [O]ptional arguments is not the biggest [problem] IMO. The biggest is that you can't do stuff like this:
>
> ```javascript
> R.compose(R.flip(R.map)([1,2,3,4]), R.add)(3);
> ```

This would be addressed by not returning a curried function.

> Attempting to automatically curry the function returned by `compose` is too naive. If a function is only guessing the correct arity half the time then it's better to not guess at all.

This would be addressed by requiring users to opt in to currying (via the new function).

It's interesting to note that if `R.add` reported its `length` as `1`—the correct length according to its documented type signature—the example would behave as you would like.

> To me the lack of a good name for the last proposed function is bad sign.

True, although this is offset by how easy it is to depict the function's behaviour in ASCII. :)

    Composition of binary functions `f` and `g`:

              +-------+
    --- a --->|       |
              |   f   |                +-------+
    --- b --->|       |--- f(a, b) --->|       |
              +-------+                |   g   |
    --- c ---------------------------->|       |--- g(f(a, b), c) --->
                                       +-------+

**Comment by CrossEye** [comment]
Although I too have a preference that `pipe` and `compose` are symmetric, it's for me only a minor consideration which could be overridden by many other factors.

The lack of a name is likewise only a small issue at the moment:  all these things, as well as the current `useWith`, `converge`, and `composeX`/`pipeX` functions are variations on a common theme.  The problem is not the dearth of names, but a surfeit of candidate uses of those names.

**Comment by CrossEye** [comment]
> It's interesting to note that if R.add reported its length as 1—the correct length according to its documented type signature—the example would behave as you would like.

I think that's a bit naive.  Even in the Haskell world, where functions really are technically unary, it's common to think of them as polyadic in many cases.

    add :: Num -> Num -> Num

is not really a function on one number that returns (Num -> Num), not in our implementation.  It is a binary function that is curried to allow you to _call_ it in that manner if you choose.  If it were really a unary function, then `add(3, 5)` would return `b => 3 + b` rather than `8`.

Our signatures admit to some ambiguity, just as they do in Haskell, and to me that's perfect, as that ambiguity exactly matches the ambiguity in the way you may call the function.  In Haskell that ambiguity is only in interpretation, function application is left-associative, so `add a b` _is_ `(add(a))(b)`.  For us the ambiguity is more fundamental.  `add :: Num -> Num -> Num` means that these are equally legitimate:  `add(a)(b)` and `add(a, b)`.  Neither, in our language, is actually just sugar for the other; it's only through our `curry` function that they're even related.

**Comment by paldepind** [comment]
@CrossEye I very much agree.

@davidchambers Oh. Now I think I get it. ASCII isn't even necessary. The type signature is sufficient: `(c -> d) -> (a -> b -> c) -> a -> b -> d`. It's the same as [Haskell's `.:`](http://hackage.haskell.org/package/composition-1.0.1.1/docs/Data-Composition.html).

Regarding variadic functions (I saw your talk about this on Gitter): If a function takes a single argument _and_ that argument is a list _and_ that list if most often typed by hand, then I think making the function variadic is a much more practical position. I don't want to type `R.compose([f, g, h])`. 

I don't think we should remove variadic function just for the sake of removing variadic functions. Normally variadic functions ruin currying so we get a benefit by removing it. But a function of the type described above isn't curry-able anyway.

**Comment by TheLudd** [comment]
@paldepind Right on about nothing gained by turning a varardic function into a unary function that takes a list as its single argument.

**Comment by CrossEye** [comment]
@paldepind:

That ASCII was just one simple example.  It's too bad that David killed off the diagrams from #1260.  Here's another sample:

    Composition of binary function `f`, ternary function `g`, and binary function `h`

              +-------+
    --- a --->|       |
              |   f   |                +-------+
    --- b --->|       |--- f(a, b) --->|       |
              +-------+                |       |                          +-------+
    --- c ---------------------------->|   g   |--- g(f(a, b), c, d) ---> |       |
                                       |       |                          |       |
    --- d ---------------------------->|       |                          |   h   |--- h(g(f(a, b), c, d), e) --->
                                       +-------+                          |       |
    --- e --------------------------------------------------------------> |       |
                                                                          +-------+


And you could add unary functions in anywhere without altering the arity or the general shape of the function.

**Comment by davidchambers** [comment]
> It's too bad that David killed off the diagrams from #1260.

The diagrams are still there, but one needs to dig into the concealed diff comments to find them. ;)

**Comment by davidchambers** [comment]
> [N]othing [is] gained by turning a varardic function into a unary function that takes a list as its single argument.

I disagree: #1213.

**Comment by CrossEye** [comment]
> The diagrams are still there,
Sure enough!  Here's another good example: https://github.com/ramda/ramda/pull/1260#discussion_r34372689

**Comment by paldepind** [comment]
> I disagree: #1213.

I don't think the difference in how the diffs looks is very significant at all.

The second issue is not a big problem IMO. We have `Function.prototype.apply` and in ECMAScript 6 invoking a function with an array is even easier. It would actually be fine with me if `compose` and `pipe` could both take a variadic number of argument _or_ an array. One would be for single line invocations and the other for multiple lines. It would make sense IMO but I know that is not generally how Ramda does things.

**Comment by davidchambers** [comment]
> I don't think the difference in how the diffs looks is very significant at all.

… and I don't think typing an extra `[` and `]` is very significant. ;)

**Comment by davidchambers** [comment]
Resolved by #1391
